File: go/day03/main.go
package main

import (
	"fmt"
	"strings"

	utils "github.com/MaraudingAvenger/Advent2025/common"
)

func FindLargestNumber(n string, d uint) uint64 {
	// Find the largest number of exactly d digits from n while preserving order.
	if len(n) < int(d) {
		panic(fmt.Sprintf("n must have at least %d digits", d))
	}
	if d == 0 {
		return 0
	}

	targetLen := int(d)
	toDrop := len(n) - targetLen
	stack := make([]byte, 0, len(n))

	for i := 0; i < len(n); i++ {
		c := n[i]
		if c < '0' || c > '9' {
			panic(fmt.Sprintf("n must contain only digits: found %q", c))
		}

		for toDrop > 0 && len(stack) > 0 && stack[len(stack)-1] < c {
			stack = stack[:len(stack)-1]
			toDrop--
		}

		stack = append(stack, c)
	}

	if toDrop > 0 {
		stack = stack[:len(stack)-toDrop]
	}

	selected := stack[:targetLen]

	var largest uint64
	for _, c := range selected {
		largest = largest*10 + uint64(c-'0')
	}

	return largest
}

func main() {
	lines, err := utils.ReadInputFile("inputs/day3.txt")
	if err != nil {
		panic(err)
	}

	var totalJolts uint64 = 0

	for _, line := range strings.Fields(lines) {
		line = strings.TrimSpace(line)
		if len(line) == 0 {
			continue
		}

		largest := FindLargestNumber(line, 2)
		totalJolts += largest
	}

	fmt.Printf("Part 1: %d\n", totalJolts)

	totalJolts = 0
	for _, line := range strings.Fields(lines) {
		line = strings.TrimSpace(line)
		if len(line) == 0 {
			continue
		}

		largest := FindLargestNumber(line, 12)
		totalJolts += largest
	}

	fmt.Printf("Part 2: %d\n", totalJolts)
}

File: rust/day03/src/main.rs
use common::read_lines;
use common::find_input_file;

fn find_largest_number(numbers: &String, n_digits: usize) -> Option<usize> {
    // find the largest number with n_digits in the string of numbers
    // the number does not need to be coniguous, but the digits must be in order
    let mut to_drop = numbers.len() - n_digits;
    let mut stack: Vec<char> = Vec::new();

    for i in 0..numbers.len() {
        let c = numbers.chars().nth(i).unwrap();
        if c < '0' || c > '9' {
            continue;
        }
        while to_drop > 0 && !stack.is_empty() && *stack.last().unwrap() < c {
            stack.pop();
            to_drop -= 1;
        }
        stack.push(c);
    }

    if to_drop > 0 {
        stack.truncate(stack.len() - to_drop);
    }
    let result = stack.iter().take(n_digits).collect::<String>();
    Some(result.parse::<usize>().unwrap())
}

fn main() {
    let lines = read_lines(&find_input_file("inputs/day3.txt"));

    let mut total_joltage: usize = 0;
    for line in &lines {
        if let Some(largest_number) = find_largest_number(line, 2) {
            total_joltage += largest_number;
        }
    }
    println!("Part 1: {}", total_joltage);

    total_joltage = 0;
    for line in &lines {
        if let Some(largest_number) = find_largest_number(line, 12) {
            total_joltage += largest_number;
        }
    }
    println!("Part 2: {}", total_joltage);
}
